src/subsystems/memory/memory_types.rs

```diff
@@ -0,0 +1,230 @@
+use crate::{crtp_utils::WaitForPacket, Error, Result};
+use crazyflie_link::Packet;
+use flume as channel;
+use std::convert::{TryFrom, TryInto};
+
+use crate::crazyflie::MEMORY_PORT;
+
+const READ_CHANNEL: u8 = 1;
+const WRITE_CHANNEL: u8 = 2;
+
+const MEM_MAX_REQUEST_SIZE: usize = 20;
+
+
+/// Description of a memory in the Crazyflie
+#[derive(Debug, Clone)]
+pub struct MemoryBackend {
+    /// Unique identifier for this memory subsystem (used when reading/writing/querying)
+    pub memory_id: u8,
+    /// Type of memory
+    pub memory_type: MemoryType,
+
+    pub(crate) uplink: channel::Sender<Packet>,
+    //TODO: Lock this!
+    pub(crate) read_downlink: channel::Receiver<Packet>,
+    pub(crate) write_downlink: channel::Receiver<Packet>,
+}
+/// Description of a memory in the Crazyflie
+#[derive(Debug, Clone)]
+pub struct MemoryDevice {
+    /// Unique identifier for this memory subsystem (used when reading/writing/querying)
+    pub memory_id: u8,
+    /// Type of memory
+    pub memory_type: MemoryType,
+    /// Size of the memory in bytes
+    pub size: u32
+}
+
+impl MemoryBackend {
+    pub(crate) async fn read(&self, address: usize, length: usize) -> Result<Vec<u8>> {
+        let mut data = vec![0; length];
+        let mut current_address = address;
+        while current_address < address + length {
+            let to_read = std::cmp::min(
+                MEM_MAX_REQUEST_SIZE,
+                (address + length) - current_address,
+            );
+            let mut request_data = Vec::new();
+            request_data.extend_from_slice(&self.memory_id.to_le_bytes());
+            request_data.extend_from_slice(&(current_address as u32).to_le_bytes());
+            request_data.extend_from_slice(&(to_read as u8).to_le_bytes());
+            let pk = Packet::new(MEMORY_PORT, READ_CHANNEL, request_data.clone());
+            self.uplink
+                .send_async(pk)
+                .await
+                .map_err(|_| Error::Disconnected)
+                .ok();
+
+            let pk = self
+                .read_downlink
+                .wait_packet(MEMORY_PORT, READ_CHANNEL, &request_data[0..5])
+                .await;
+            if let Ok(pk) = pk {
+                let pk_data = pk.get_data();
+                let read_address = u32::from_le_bytes(pk_data[1..5].try_into().unwrap());
```

Comment:
In Python lib `cflib/crazyflie/mem/__init__.py`  `_handle_chan_read`
```
        (addr, status) = struct.unpack('<IB', payload[0:5])
        data = struct.unpack('B' * len(payload[5:]), payload[5:])
```

So 1 to 4 address and 5 a status byte?

Reply:
Good catch, fixed.